### ISSUE TITLE
Memorization task input and target 

### ISSUE THREAD

**Issue by AjayTalati** [wojciechz/learning_to_execute#5]
Hi,

I'm trying to reproduce the results from the paper for the sequence memorization task. May I ask how you constructed the input and target sequences, for the doubling and inverting strategies?

Say for example the task is to memorize/reconstruct the sequence `abcd`. I've tried to reproduce your results for the different curriculum strategies using, the input `x` and target `y`,

`x = <sos>abcd<eos><sos>dcba<eos><sos>abcd<eos>000000`

`y = 000000000000000000<sos>abcd<eos>`

So I've added start and end of sequence tokens, `<sos>` and `<eos>` and I present the sequence to be memorized, then its inversion and then it's double, followed by padding zeros in the input, `x`. Thus the target `y` in this example has 18 zeros, (3 x 6), before the sequence to be memorized is presented. 

May I ask if this is how you constructed the data for the memorization task in your paper?

Kind regards,

Aj 

PS - I also have used a `nn.ClassNLLCriterion` criterion after the mask, so maybe this has interfered with reproducing the results in the paper?

**Comment by wojzaremba** [comment]
1. Let's say that strategy is double, then data would look like:

x = <sos>abcd<eos><sos>abcs<eos>

y = <sos>abcd<eos>

2. Let's say that strategy is invert, then data would look like:

x = <sos>dcba<eos>

y = <sos>abcd<eos>

3. Let's say that strategy is double and invert, then data would look like

x = <sos>dcba<eos><sos>dcba<eos>

y = <sos>abcd<eos>

**Comment by AjayTalati** [comment]
Ha ha! That's really cute :+1:

Thank you very much, for taking the time to answer :)

I'll do as you state - and calculate with this data - my implementation is quite different though? 

I'll try and go back to looking at your code, the `MaskedLoss` module and how the accumulated loss/score is calculated. Guess that's where I've made an error?

PS - maybe when you have a few minuets spare you might be able to comment on how your code works - it's not a big deal though - I should be able to replicate your results using my own implementation. 

I don't really understand why the returned [accumulated loss](https://github.com/wojciechz/learning_to_execute/blob/master/main.lua#L151) is `0` in `eval_training`? 

May I ask how the score/loss is calculated in the code? If I understand correctly, the `MaskedLoss` module returns `1` as it's second output if the class with the highest log probability is the same as the target class - lines [30 and 31](https://github.com/wojciechz/learning_to_execute/blob/master/layers/MaskedLoss.lua#L30). This is then normalized in `fp` in lines [121 to 124](https://github.com/wojciechz/learning_to_execute/blob/master/main.lua#L121). I'm not sure how SGD maximizes this score, or how it knows what value `state_train.acc` is? 

I could'nt really understand this so I've tried to build my own implementation using class cross entropy, it seems to work fairly well, but it's not very efficient due to how the mask is implemented :(
